text/0517-io-os-reform.md

```diff
@@ -708,7 +712,129 @@ throughout IO, we can go on to explore the modules in detail.
 ### `std::fs`
 [std::fs]: #stdfs
 
-> To be added in a follow-up PR.
+The `fs` module will provide most of the functionality it does today,
+but with a stronger cross-platform orientation.
+
+Note that all path-consuming functions will now take an
+`AsPath`-bounded parameter for ergonomic reasons (this will allow
+passing in Rust strings and literals directly, for example).
+
+#### Free functions
+[Free functions]: #free-functions
+
+**Files**:
+
+* `copy`. Take `AsPath` bound.
+* `rename`. Take `AsPath` bound.
+* `remove_file` (renamed from `unlink`). Take `AsPath` bound.
+
+* `file_attr` (renamed from `stat`). Take `AsPath` bound. Yield a new
+  struct, `FileAttr`, with no public fields, but `size`, `kind` and
+  `perm` accessors. The various `os::platform` modules will offer
+  extension methods on this structure.
+
+* `set_perm` (renamed from `chmod`). Take `AsPath` bound, and a
+  `FilePermissions` value. The `FilePermissions` type will be revamped
+  as a struct with private implementation; see below.
+
+**Directories**:
+
+* `make_dir` (renamed from `mkdir`). Take `AsPath` bound.
+* `make_dir_all` (renamed from `mkdir_recursive`). Take `AsPath` bound.
+* `read_dir` (renamed from `readdir`). Take `AsPath` bound. Yield a
+  newtypes iterator, which yields a new type `DirEntry` which has an
+  accessor for `Path`, but will eventually provide other information
+  as well (possibly via platform-specific extensions).
+* `remove_dir` (renamed from `rmdir`). Take `AsPath` bound.
+* `remove_dir_all` (renamed from `rmdir_recursive`). Take
+  `AsPath` bound.
+* `walk_dir`. Take `AsPath` bound. Yield an iterator over `IoResult<DirEntry>`.
+
+**Links**:
+
+* `hard_link` (renamed from `link`). Take `AsPath` bound.
+* `sym_link` (renamed from `symlink`). Take `AsPath` bound.
+* `read_link` (renamed form `readlink`). Take `AsPath` bound.
+
+#### Files
+[Files]: #files
+
+The `File` type will largely stay as it is today, except that it will
+use the `AsPath` bound everywhere.
```

Comment:
to be clear, this is what bothers me:

    let p = Path::new("/dev/null");
    let f = File::open(&p).unwrap().as_raw_fd();
    println!("fd: {}", f);

That's a dangling, non-sensical "pointer" even in single threaded, 100%
safe code.  Can as_raw_fd just be made unsafe?

On Mon, Jan 26, 2015 at 7:42 PM, Alex Crichton <notifications@github.com>
wrote:

> In text/0517-io-os-reform.md
> <https://github.com/rust-lang/rfcs/pull/739#discussion_r23581430>:
>
> > +* `remove_dir` (renamed from `rmdir`). Take `AsPath` bound.
> > +* `remove_dir_all` (renamed from `rmdir_recursive`). Take
> > +  `AsPath` bound.
> > +* `walk_dir`. Take `AsPath` bound. Yield an iterator over `IoResult<DirEntry>`.
> > +
> > +**Links**:
> > +
> > +* `hard_link` (renamed from `link`). Take `AsPath` bound.
> > +* `sym_link` (renamed from `symlink`). Take `AsPath` bound.
> > +* `read_link` (renamed form `readlink`). Take `AsPath` bound.
> > +
> > +#### Files
> > +[Files]: #files
> > +
> > +The `File` type will largely stay as it is today, except that it will
> > +use the `AsPath` bound everywhere.
>
> I consider our story here very analagous to that of plain old memory
> management pointers like Box<T> You can get a very unsafe *mut T, for
> example, safely from a Box<T> but mutation is all unsafe (similar to how
> as_raw_fd is safe but usage is likely unsafe). Similarly, Box<T> doesn't
> attempt to check if the memory is valid and I don't think that our bindings
> will special-case EBADF as well.
>
> In general each object will own its file descriptor. The descriptor can be
> inspected and lent out via a safe API (but in this case platform specific)
> and relies on modification being unsafe elsewhere. You are, however,
> guaranteed that the raw component (in this case a file descriptor) is valid
> so long as the file itself is still alive.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rfcs/pull/739/files#r23581430>.
>